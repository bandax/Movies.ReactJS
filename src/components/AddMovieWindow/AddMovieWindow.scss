.add-movie-window {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  transition: opacity 500ms;
}

.add-movie-modal {
  margin: 10px 70px 0 70px;
  padding: 20px;
  background: #000000;
  border-radius: 5px;
  width: 70%;
  position: relative;
  transition: all 5s ease-in-out;
}

.add-movie-title {
  margin: 10px 0 20px 0;
  color: #ffffff;
  text-transform: uppercase;
  font-size: 35px;
  font-weight: bold;
}

.add-movie-close {
  background-color: transparent;
  position: absolute;
  top: 5px;
  right: 30px;
  transition: all 200ms;
  font-size: 30px;
  font-weight: bold;
  text-decoration: none;
  color: #ffffff;
  font-size: 40px;
  border: none;
  cursor: pointer;
}

.add-movie-content {
  max-height: 30%;
  overflow: auto;
}

.react-datepicker-wrapper {
  width: 100%;
}

// Forms controls
.input-text {
  width: 96%;
  padding: 20px 20px;
  margin: 15px 0 10px 10px;
  display: inline-block;
  border: 1px solid #232323;
  border-radius: 4px;
  box-sizing: border-box;
  background-color: rgba(10, 10, 10, 0.8);
  color: #ffffff;
  font-size: 18px;
}

.date-icon {
  background: rgba(10, 10, 10, 0.8) url("../../../assets/date_icon.png")
    no-repeat;
  background-position: 99% 50%;
  background-size: 30px 29px;
  cursor: pointer;
}

.input-select {
  width: 95%;
  padding: 20px 20px;
  margin: 15px 0 10px 10px;
  display: inline-block;
  border: 1px solid #232323;
  border-radius: 4px;
  box-sizing: border-box;
  background-color: rgba(10, 10, 10, 0.8);
  color: #ffffff;
  font-size: 18px;
}

.label-text {
  color: #f65261;
  text-transform: uppercase;
  font-weight: bold;
  padding: 0 0 10px 0;
  font-size: 18px;
}

.btn-submit {
  background-color: #f65261;
  color: white;
  font-size: 18px;
  padding: 10px 50px;
  border-radius: 5px;
  text-transform: uppercase;
}

.btn-reset {
  background-color: #232323;
  color: #f65261;
  font-size: 18px;
  padding: 10px 50px;
  border-radius: 5px;
  text-transform: uppercase;
}

.multi-select {
  --rmsc-main: #4285f4;
  --rmsc-hover: #f1f3f5;
  --rmsc-selected: #e2e6ea;
  --rmsc-border: #ccc;
  --rmsc-gray: #aaa;
  --rmsc-bg: rgba(10, 10, 10, 0.8) !important;
  --rmsc-p: 10px; // Spacing
  --rmsc-radius: 4px; // Radius
  --rmsc-h: 38px; // Height
  --rmsc-w: 95% !important; // Width
}

.dropdown-container {
  width: 96%;
  border: 1px solid #232323;
  border-radius: 4px;
  font-size: 18px;
  padding: 10px 20px;
  margin: 15px 0 10px 10px;
}

@media screen and (max-width: 700px) {
  .box {
    width: 70%;
  }
  .add-movie-modal {
    width: 70%;
  }
}
