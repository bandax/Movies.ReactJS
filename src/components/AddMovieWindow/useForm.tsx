import * as React from "react";

const useForm = (initialValues: any) => {
  const [values, setValues] = React.useState(initialValues);

  return [
    values,
    (e: any, initialValues: any) => {
      if (!initialValues) {
        setValues({
          ...values,
          [e.currentTarget.name]: e.currentTarget.value,
        });
      } else {
        setValues(initialValues);
      }
    },
  ];
};

export { useForm };
