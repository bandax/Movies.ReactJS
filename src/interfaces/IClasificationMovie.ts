export interface IClasification {
    id: string;
    name: string;
}
