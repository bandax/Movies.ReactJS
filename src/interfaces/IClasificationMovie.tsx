export interface IClasification {
    id: string,
    name: string
}

export interface IClasificationMovieProps {
    clasificationTypes: IClasification[],    
}

