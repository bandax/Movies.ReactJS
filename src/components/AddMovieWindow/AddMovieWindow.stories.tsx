import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import clasificationTypes from '../../data/clasifications.json';

import { AddMovieWindow, IAddMovieWindowProps } from './AddMovieWindow';

export default {
  title: 'Movies/Add Movie Window',
  component: AddMovieWindow,
} as Meta;

const Template: Story<IAddMovieWindowProps> = (args) => (
  <AddMovieWindow {...args} />
);

export const HideWindow = Template.bind({});
HideWindow.args = {
  showModal: false,
  clasificationMovies: clasificationTypes,
};

export const ShowWindow = Template.bind({});
ShowWindow.args = {
  showModal: true,
  clasificationMovies: clasificationTypes,
};
