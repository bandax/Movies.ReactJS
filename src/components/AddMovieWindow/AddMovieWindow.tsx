import * as React from 'react';
import './AddMovieWindow.scss';
import { IClasification } from '../../interfaces/IClasificationMovie';
import { IMovieData } from '../../interfaces/IMovieData';
import * as Yup from 'yup';
import { Formik, Form } from 'formik';
import { FormikTextInput } from '../Shared/FormikTextInput';
import { FormikMultiSelect } from '../Shared/FormikMultiSelect';
import { FormikDatePicker } from '../Shared/FormikDatePicker';
import { IOption } from '../../interfaces/IOption';

interface IAddMovieWindowProps {
  showModal: boolean;
  clasificationMovies: IClasification[];
  movie?: IMovieData;
  onShowAddMovieWindow: () => void;
  onAddMovieSubmit: (movie: IMovieData) => void;
}

interface IFormMovieValues {
  id: number;
  title: string;
  budget: number;
  revenue: number;
  tagline: string;
  vote_average: number;
  vote_count: number;
  overview: string;
  poster_path: string;
  runtime: number;
  release_date: Date;
  genres: IOption[];
}

const AddMovieWindow: React.FunctionComponent<IAddMovieWindowProps> = (
  props: IAddMovieWindowProps
) => {
  const [genreOptions, setOptionGenres] = React.useState([]);

  React.useEffect(() => {
    const optGenres: IOption[] = props.clasificationMovies.map(
      (option: IClasification) => ({
        label: option.name,
        value: option.id,
      })
    );

    setOptionGenres(optGenres);
  }, [props.clasificationMovies]);

  const [selectedGenres, setSelectedGenres] = React.useState([]);

  React.useEffect(() => {
    const selGenres: IOption[] = [];
    props.movie?.genres.forEach((genre) => {
      const [selectedGenre] = genreOptions.filter(
        (g: IOption) => g.label === genre
      );
      if (selectedGenre) {
        selGenres.push(selectedGenre);
      }
    });
    setSelectedGenres(selGenres);
  }, [genreOptions]);

  const initialValues: IFormMovieValues = {
    budget: props.movie?.budget ?? 0,
    revenue: props.movie?.revenue ?? 0,
    tagline: props.movie?.tagline ?? '',
    vote_average: props.movie?.vote_average ?? 0,
    vote_count: props.movie?.vote_count ?? 0,
    id: props.movie?.id ?? 0,
    title: props.movie?.title ?? '',
    overview: props.movie?.overview ?? '',
    poster_path: props.movie?.poster_path ?? '',
    runtime: props.movie?.runtime ?? 0,
    release_date: props.movie?.release_date
      ? new Date(props.movie?.release_date)
      : new Date(),
    genres: selectedGenres,
  };

  const validationSchema = Yup.object({
    title: Yup.string().required('Required'),
    runtime: Yup.number().required('Required').positive().integer(),
    budget: Yup.number().required('Required').positive().integer(),
    revenue: Yup.number().required('Required').positive().integer(),
    vote_average: Yup.number().required('Required'),
    vote_count: Yup.number().required('Required').positive().integer(),
    overview: Yup.string().required('Required'),
    tagline: Yup.string().required('Required'),
    poster_path: Yup.string().required('Required').url(),
    release_date: Yup.date().default(() => new Date()),
  });

  const showMovieIdField = () => {
    return initialValues.id !== 0 ? (
      <>
        <FormikTextInput
          id="id"
          name="id"
          label="Movie Id"
          placeholder=""
          type="text"
          readOnly="{true}"
        />
      </>
    ) : null;
  };

  if (!props.showModal) {
    return null;
  }

  return (
    <div className="add-movie-window">
      <div className="add-movie-modal">
        <h2 className="add-movie-title">Add movie window</h2>
        <button
          className="add-movie-close"
          data-testid="add-movie-close"
          onClick={props.onShowAddMovieWindow}
        >
          &times;
        </button>
        <Formik
          initialValues={initialValues}
          validationSchema={validationSchema}
          onSubmit={(values, actions) => {
            const selGenres: string[] = values.genres.map((option: IOption) => {
              return option.label;
            });
            console.log(values.release_date);

            const movieToSave: IMovieData = {
              ...values,
              genres: selGenres,
              release_date: values.release_date.toLocaleDateString(),
            };

            props.onAddMovieSubmit(movieToSave);
          }}
        >
          <Form>
            <div className="add-movie-content">
              <div className="row">
                <div className="col-6">
                  {showMovieIdField()}

                  <FormikTextInput
                    id="title"
                    name="title"
                    label="Title"
                    placeholder="Title"
                    type="text"
                  />

                  <FormikDatePicker
                    id="release_date"
                    name="release_date"
                    label="Release Date"
                  />

                  <FormikTextInput
                    id="poster_path"
                    name="poster_path"
                    label="Movie URL"
                    placeholder="Movie URL here"
                    type="text"
                  />

                  <FormikMultiSelect
                    id="genres"
                    name="genres"
                    label="Genre"
                    selValue={selectedGenres}
                    options={genreOptions}
                  />

                  <FormikTextInput
                    id="overview"
                    name="overview"
                    label="Overview"
                    placeholder="Overview here"
                    type="text"
                  />
                </div>
                <div className="col-6">
                  <FormikTextInput
                    id="runtime"
                    name="runtime"
                    label="Runtime"
                    placeholder="Runtime here"
                    type="number"
                  />

                  <FormikTextInput
                    id="budget"
                    name="budget"
                    label="Budget"
                    placeholder="Budget"
                    type="number"
                  />

                  <FormikTextInput
                    id="tagline"
                    name="tagline"
                    label="Tagline"
                    placeholder="Tagline"
                    type="text"
                  />

                  <FormikTextInput
                    id="revenue"
                    name="revenue"
                    label="Revenue"
                    placeholder="Revenue here"
                    type="number"
                  />

                  <FormikTextInput
                    id="vote_average"
                    name="vote_average"
                    label="Vote Avg"
                    placeholder="vote"
                    type="number"
                  />

                  <FormikTextInput
                    id="vote_count"
                    name="vote_count"
                    label="Vote Count"
                    placeholder="vote count"
                    type="number"
                  />
                </div>
              </div>
            </div>
            <div className="actions">
              <button className="btn btn-reset" type="reset">
                Reset
              </button>
              <button className="btn btn-submit" type="submit">
                Submit
              </button>
            </div>
          </Form>
        </Formik>
      </div>
    </div>
  );
};

export { AddMovieWindow };
