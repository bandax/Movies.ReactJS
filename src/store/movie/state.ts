import { IMovie } from "../../interfaces/IResultsMovies";

export interface MovieState {
  movies: IMovie[];
  loading: boolean;
  errorMessage: string;
}
