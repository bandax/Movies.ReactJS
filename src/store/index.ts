import { combineReducers, applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { movieReducer } from './movie/reducers';

const rootReducer = combineReducers({
  movie: movieReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export const store = createStore(rootReducer, applyMiddleware(thunk));
