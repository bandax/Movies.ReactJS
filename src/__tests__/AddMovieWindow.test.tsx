import * as React from 'react';
import { AddMovieWindow } from '../components/AddMovieWindow/AddMovieWindow';
import thunk, { ThunkDispatch } from 'redux-thunk';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import configureMockStore, { MockStoreCreator } from 'redux-mock-store';
import { MovieState } from '../store/movie/state';
import { Action } from 'redux';
import { Provider } from 'react-redux';
import { RootState } from '../store/index';
import clasificationTypes from '../data/clasifications.json';
import { IMovieData } from '../interfaces/IMovieData';

type DispatchExts = ThunkDispatch<MovieState, void, Action>;
const middleware = [thunk];
const state: RootState = {
  movie: {
    movie: null,
    errorMessage: '',
    findMovies: true,
    loading: false,
    movies: [],
  },
};

const mockStore = configureMockStore<RootState, DispatchExts>(middleware);

describe('<AddMovie />', () => {
  let store;

  const newMovie: IMovieData = {
    id: 0,
    title: 'Avengers Infinity War',
    poster_path: '../../../assets/posters/avengers-infinity-war.PNG',
    release_date: '01/02/2010',
    budget: 123,
    revenue: 123,
    vote_average: 123,
    vote_count: 123,
    runtime: 412,
    tagline: '12343',
    overview: 'tests',
    genres: ['Action & Adventure', 'Comedy'],
  };

  const movieToUpdate: IMovieData = {
    id: 1234,
    title: 'Avengers Infinity War',
    poster_path: '../../../assets/posters/avengers-infinity-war.PNG',
    release_date: '01/02/2010',
    budget: 123,
    revenue: 123,
    vote_average: 123,
    vote_count: 123,
    runtime: 412,
    tagline: '12343',
    overview: 'tests',
    genres: ['Action & Adventure', 'Comedy'],
  };

  let onAddMovieSubmit;
  let onShowAddMovieWindow;

  beforeEach(() => {
    onAddMovieSubmit = jest.fn();
    onShowAddMovieWindow = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should render a Add Movie Window Component', async () => {
    const { asFragment } = render(
      <AddMovieWindow
        clasificationMovies={clasificationTypes}
        onAddMovieSubmit={onAddMovieSubmit}
        onShowAddMovieWindow={onShowAddMovieWindow}
        showModal={true}
        movie={newMovie}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should not render the component at all', async () => {
    const { container } = render(
      <AddMovieWindow
        clasificationMovies={clasificationTypes}
        onAddMovieSubmit={onAddMovieSubmit}
        onShowAddMovieWindow={onShowAddMovieWindow}
        showModal={false}
        movie={newMovie}
      />
    );
    expect(container.firstChild).toBeNull();
  });

  test('should close add movie window when click close icon', async () => {
    const { getByTestId, getByText } = render(
      <AddMovieWindow
        clasificationMovies={clasificationTypes}
        onAddMovieSubmit={onAddMovieSubmit}
        onShowAddMovieWindow={onShowAddMovieWindow}
        showModal={true}
        movie={newMovie}
      />
    );
    const button = getByTestId('add-movie-close');
    userEvent.click(button);
    expect(getByText('Add movie window')).toBeInTheDocument();
    await waitFor(() => {
      expect(onShowAddMovieWindow).toHaveBeenCalled();
    });
  });

  test('should show moveId input control', async () => {
    const { getByTestId, getByText } = render(
      <AddMovieWindow
        clasificationMovies={clasificationTypes}
        onAddMovieSubmit={onAddMovieSubmit}
        onShowAddMovieWindow={onShowAddMovieWindow}
        showModal={true}
        movie={movieToUpdate}
      />
    );
    const input = getByText('Movie Id');
    expect(input).toBeInTheDocument();
  });

  test('should allow to call submit button', async () => {
    const { getByTestId, getByText } = render(
      <AddMovieWindow
        clasificationMovies={clasificationTypes}
        onAddMovieSubmit={onAddMovieSubmit}
        onShowAddMovieWindow={onShowAddMovieWindow}
        showModal={true}
        movie={null}
      />
    );

    userEvent.type(getByText('Title'), 'New Movie');
    userEvent.type(
      getByText('Movie URL'),
      'https://upload.wikimedia.org/wikipedia/commons/thumb/d/d7/Scarymovie4-logo.svg/1920px-Scarymovie4-logo.svg.png'
    );
    userEvent.type(getByText('Overview'), 'Overview');
    userEvent.type(getByText('Runtime'), '123');
    userEvent.type(getByText('Budget'), '123');
    userEvent.type(getByText('Tagline'), '123');
    userEvent.type(getByText('Revenue'), '1234');
    userEvent.type(getByText('Vote Avg'), '1234');
    userEvent.type(getByText('Vote Count'), '23454');

    const button = getByText(/Submit/i);
    userEvent.click(button);

    await waitFor(() => {
      expect(onAddMovieSubmit).toHaveBeenCalled();
    });
  });

  test('should allow to call reset button and clean up values', async () => {
    const { getByTestId, getByText } = render(
      <AddMovieWindow
        clasificationMovies={clasificationTypes}
        onAddMovieSubmit={onAddMovieSubmit}
        onShowAddMovieWindow={onShowAddMovieWindow}
        showModal={true}
        movie={null}
      />
    );

    const title = getByText('Title');
    userEvent.type(title, 'New Movie');
    const button = getByText(/Reset/i);
    userEvent.click(button);

    expect(title).not.toHaveValue();
  });
});
