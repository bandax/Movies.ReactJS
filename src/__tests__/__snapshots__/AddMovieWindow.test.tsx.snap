// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddMovie /> should render a Add Movie Window Component 1`] = `
<DocumentFragment>
  <div
    class="add-movie-window"
  >
    <div
      class="add-movie-modal"
    >
      <h2
        class="add-movie-title"
      >
        Add movie window
      </h2>
      <button
        class="add-movie-close"
        data-testid="add-movie-close"
      >
        Ã—
      </button>
      <form
        action="#"
      >
        <div
          class="add-movie-content"
        >
          <div
            class="row"
          >
            <div
              class="col-6"
            >
              <label
                class="label-text"
                for="title"
              >
                Title
              </label>
              <input
                class="input-text"
                id="title"
                label="Title"
                name="title"
                placeholder="Title"
                type="text"
                value="Avengers Infinity War"
              />
              <label
                class="label-text"
                for="release_date"
              >
                Release Date
              </label>
              <div
                class="react-datepicker-wrapper"
              >
                <div
                  class="react-datepicker__input-container"
                >
                  <input
                    class="date-icon input-text"
                    id="release_date"
                    name="release_date"
                    type="text"
                    value="01/02/2010"
                  />
                </div>
              </div>
              <label
                class="label-text"
                for="poster_path"
              >
                Movie URL
              </label>
              <input
                class="input-text"
                id="poster_path"
                label="Movie URL"
                name="poster_path"
                placeholder="Movie URL here"
                type="text"
                value="../../../assets/posters/avengers-infinity-war.PNG"
              />
              <label
                class="label-text"
                for="genres"
              >
                Genre
              </label>
              <div
                class="go2139093995 multi-select"
              >
                <div
                  aria-expanded="false"
                  aria-labelledby="Select"
                  aria-readonly="true"
                  class="go2646822163 dropdown-container"
                  tabindex="0"
                >
                  <div
                    class="go2642161244 dropdown-heading"
                  >
                    <div
                      class="dropdown-heading-value"
                    >
                      <span
                        class="gray"
                      >
                        Select...
                      </span>
                    </div>
                    <svg
                      class="dropdown-heading-dropdown-arrow gray"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-width="2"
                      width="24"
                    >
                      <path
                        d="M6 9L12 15 18 9"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <label
                class="label-text"
                for="overview"
              >
                Overview
              </label>
              <input
                class="input-text"
                id="overview"
                label="Overview"
                name="overview"
                placeholder="Overview here"
                type="text"
                value="tests"
              />
            </div>
            <div
              class="col-6"
            >
              <label
                class="label-text"
                for="runtime"
              >
                Runtime
              </label>
              <input
                class="input-text"
                id="runtime"
                label="Runtime"
                name="runtime"
                placeholder="Runtime here"
                type="number"
                value="412"
              />
              <label
                class="label-text"
                for="budget"
              >
                Budget
              </label>
              <input
                class="input-text"
                id="budget"
                label="Budget"
                name="budget"
                placeholder="Budget"
                type="number"
                value="123"
              />
              <label
                class="label-text"
                for="tagline"
              >
                Tagline
              </label>
              <input
                class="input-text"
                id="tagline"
                label="Tagline"
                name="tagline"
                placeholder="Tagline"
                type="text"
                value="12343"
              />
              <label
                class="label-text"
                for="revenue"
              >
                Revenue
              </label>
              <input
                class="input-text"
                id="revenue"
                label="Revenue"
                name="revenue"
                placeholder="Revenue here"
                type="number"
                value="123"
              />
              <label
                class="label-text"
                for="vote_average"
              >
                Vote Avg
              </label>
              <input
                class="input-text"
                id="vote_average"
                label="Vote Avg"
                name="vote_average"
                placeholder="vote"
                type="number"
                value="123"
              />
              <label
                class="label-text"
                for="vote_count"
              >
                Vote Count
              </label>
              <input
                class="input-text"
                id="vote_count"
                label="Vote Count"
                name="vote_count"
                placeholder="vote count"
                type="number"
                value="123"
              />
            </div>
          </div>
        </div>
        <div
          class="actions"
        >
          <button
            class="btn btn-reset"
            type="reset"
          >
            Reset
          </button>
          <button
            class="btn btn-submit"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
